lib/googleauth/default_credentials.rb

```diff
@@ -0,0 +1,125 @@
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require 'multi_json'
+require 'stringio'
+
+require 'googleauth/credentials_loader'
+require 'googleauth/client_id'
+require 'googleauth/compute_engine'
+require 'googleauth/service_account'
+require 'googleauth/user_authorizer'
+require 'googleauth/user_refresh'
+require 'googleauth/web_user_authorizer'
+
+module Google
+  # Module Auth provides classes that provide Google-specific authorization
+  # used to access Google APIs.
+  module Auth
+    NOT_FOUND_ERROR = <<END.freeze
+Could not load the default credentials. Browse to
+https://developers.google.com/accounts/docs/application-default-credentials
+for more information
+END
+
+    # DefaultCredentials is used to preload the credentials file, to determine
+    # which type of credentials should be loaded.
+    class DefaultCredentials
+      extend CredentialsLoader
+
+      # override CredentialsLoader#make_creds to use the class determined by
+      # loading the json.
+      def self.make_creds(options = {})
+        json_key_io, scope = options.values_at(:json_key_io, :scope)
+        if json_key_io
+          json_key, clz = determine_creds_class(json_key_io)
+          clz.make_creds(json_key_io: StringIO.new(MultiJson.dump(json_key)),
+                         scope: scope)
+        else
+          clz = read_creds
+          clz.make_creds(scope: scope)
+        end
+      end
+
+      def self.read_creds
+        env_var = CredentialsLoader::ACCOUNT_TYPE_VAR
+        type = ENV[env_var]
+        raise "#{env_var} is undefined in env" unless type
+        case type
+        when 'service_account'
+          ServiceAccountCredentials
+        when 'authorized_user'
+          UserRefreshCredentials
+        else
+          raise "credentials type '#{type}' is not supported"
+        end
+      end
+
+      # Reads the input json and determines which creds class to use.
+      def self.determine_creds_class(json_key_io)
+        json_key = MultiJson.load(json_key_io.read)
+        key = 'type'
+        raise "the json is missing the '#{key}' field" unless json_key.key?(key)
+        type = json_key[key]
+        case type
+        when 'service_account'
+          [json_key, ServiceAccountCredentials]
+        when 'authorized_user'
+          [json_key, UserRefreshCredentials]
+        else
+          raise "credentials type '#{type}' is not supported"
+        end
+      end
+    end
+
+    # Obtains the default credentials implementation to use in this
```

Comment:
It's more idiomatic to have this method stay in `googleauth.rb` since it's not part of the `DefaultCredentials` class.

Reply:
Agreed, What I am trying to avoid is a circular dependency on the `googleauth.rb` file. The credentials class I am introducing depends on the `get_application_default` method. Would moving that method into it's own file be more idiomatic?

Reply:
Oh, I see. Ick. Yes, move it to its own file.

Reply:
Done.